[Macros] Implement support for function body macros on closures.

hborla opened the pull request:
This change adds support for function body macros on closure expressions, gated behind the `ClosureBodyMacro` experimental feature.

This is pitched at https://forums.swift.org/t/pre-pitch-closure-body-macros/78534. The implementation does not match the current pitch for type checking behavior of macro arguments and expansions; I'll make that change in a follow up PR.

hborla committed:
[Macros] Support function body macros on closures.

xedin committed:
[Macros] A macro attached to a closure body covers a range after `in` and up to `}`

The macro is not allowed to write explicit type or `in` keyword.

xedin committed:
[PreCheckTarget] Move closure body expansion into `PreCheckTarget::walkToClosureExprPre`

If closure body is expanded in expression context we need to do
some special handling to make sure that source ranges are not
completely broken.

xedin committed:
[TypeCheckMacros] Never associate MacroExpansionExpr with a closure

The closures are type-checked after macros are expanded
which means that macro cannot reference any declarations
from inner or outer closures as its arguments.

For example:

`_: (Int) -> Void = { x in { @Macro(x) in ... }() }`

`x` is not going to be type-checked at macro expansion
time and cannot be referenced by `@Macro`.

Let's walk up declaration contexts until we find first
non-closure one. This means that we can support a local
declaration that is defined inside of a closure because
they are separately checked after outer ones are already
processed.

hborla committed:
[Macros] Gate closure body macros behind an experimental feature flag.

hborla commented:
https://github.com/swiftlang/swift-syntax/pull/3016

@swift-ci please smoke test

hborla commented:
https://github.com/swiftlang/swift-syntax/pull/3016

@swift-ci please smoke test

hborla committed:
[Macros] Look for specific syntax node kinds instead of `Syntax` when
expanding an attached macro.

This fixes a crash in one of the attached macro tests.

hborla commented:
@swift-ci please smoke test

hborla commented on lib/Sema/TypeCheckMacros.cpp:
```diff
@@ -1568,6 +1581,142 @@ static SourceFile *evaluateAttachedMacro(MacroDecl *macro, Decl *attachedTo,
 }
 
+static SourceFile *evaluateAttachedMacro(MacroDecl *macro,
+                                         ClosureExpr *attachedTo,
```
Much of this code is duplicated from the other overload of `evaluateAttachedMacro` that accepts a `Decl *`, but it's probably not hard to generalize using `ASTNode` instead.

rintaro commented:
Test failures
```
  Swift(linux-x86_64) :: SourceKit/Macros/diags.swift
  Swift(linux-x86_64) :: SourceKit/Macros/syntactic_expansion.swift
```
are due to extra outputs were detected, e.g.
```
unable to find node: keyword(_CompilerSwiftSyntax.Keyword.struct)
```
This is from `findSyntaxNodeInSourceFile` because you're checking `ClosureSyntax` first, then falling back to `DeclSyntax`. This happened only in Linux/WIndows because macOS uses XPC service which doesn't propagate stdout/stderr output to the host, but Linux/Windows use in-proc source kit.

hborla commented:
@rintaro ah, thank you! I think those print statements are problematic, because no matter what I check first, it's expected for the result to be `nil` in some cases. I wonder if the print statements should be removed?

hborla committed:
[Macros] Remove a `print` statement from `findSyntaxNodeInSourceFile`.

This code path isn't on an error path, and the print statement causes
unnecessary build output, which also causes test failures.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

rintaro commented on lib/ASTGen/Sources/MacroEvaluation/Macros.swift:
```diff
@@ -583,15 +583,22 @@ func expandAttachedMacro(
+  // attribute is attached.
+  let node: Syntax
+  if let closureNode = findNode(type: ClosureExprSyntax.self) {
+    node = Syntax(closureNode)
```
This may find outer unrelated `ClosureSyntax`. E.g.
```swift
_ = { (arg: Int) in
  @myBodyMacro func foo() -> Void
}
```
In this case the source location would be the `func` which is inside the closure. We need additional check, or maybe `findSyntaxNodeInSourceFile` should accept a check function like:

```swift
findSyntaxNodeInSourceFile(
  sourceFilePtr: declarationSourceFilePtr,
  sourceLocationPtr: declarationSourceLocPointer,
  where: { $0.is(DeclSyntax.self) || $0.is(ClosureExprSyntax.self) }
)

hborla replied:
Good catch, thank you! I like your idea to add an overload of `findSyntaxNodeInSourceFile` that accepts a predicate, I'll make that change.

hborla committed:
[Macros] Add an overload of `findSyntaxNodeInSourceFile` that accepts
a predicate.

The new overload is used during macro expansion to find the nearest
syntax node that is either a `DeclSyntax` or a `ClosureExprSyntax`.
This avoids an issue where calling the overload that accepts a
specific type may find an outer, unrelated declaration or closure.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test Linux

hborla commented:
@swift-ci please test Windows
